### PR TITLE
[feature] Added server side logging

### LINKED ISSUE #82: [wifi-login-pages] Add server side logging (nodejs)
Add a proper logger to the server implementation.
Log HTTP requests to openwisp-radius with INFO.
When there is an error, ensure the logging works also if the request is canceled (eg: timeout). Right now in case of timeout the request logging fails and there's no logging.

### PR DESCRIPTION
**Pull request by codesankalp**
Closes #82 

Tasks done:
1. Added Winston for logging
2. Integrated morgan with Winston to log HTTP requests
3. Removed checking for status code to handle timeout error and connection refused error.

### CODE REVIEW

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I think by default we should do:

- log level: `debug` if development, `warning` if production (as you did) but should be possible to override this in some way (env var? what's the best practice here?)
- output: console by default, but should be possible to configure it (following the same method that we'l chose for log level) to a different transport

Can we log HTTP requests by default during dev?

Please also add information to the README on how to override logging configuration.

File: server/morganMiddleware.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import Logger from "./utils/logger";
+
+const stream = {
+  write: (message) => Logger.http(message),
+};
+
+const skip = () => {
+  const env = process.env.NODE_ENV || "development";
+  return env !== "development";
```
[the remark this answers is not in the document]
Comment (codesankalp) [reply server/morganMiddleware.js:11]: Yes

File: server/loggerConfig.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const logFilePath = {
+  all: process.env.ALL_LOG_FILE || "logs/all.log",
+  error: process.env.ERROR_LOG_FILE || "logs/error.log",
+  warn: process.env.WARN_LOG_FILE || "logs/warn.log",
+  info: process.env.INFO_LOG_FILE || "logs/info.log",
+  http: process.env.HTTP_LOG_FILE || "logs/http.log",
+  debug: process.env.DEBUG_LOG_FILE || "logs/debug.log",
```
[the remark this answers is not in the document]
Comment (codesankalp) [reply server/loggerConfig.js:7]: There is a single file for all logs i.e. `logs/all.log` but I thought having separate log files for different types will be better.
This is configurable like if we want to change the log file path then we have to pass it in environment variables.
I will document these things in README.

File: server/utils/logger.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  warn: 1,
+  info: 2,
+  http: 3,
+  debug: 4,
+};
+
+const level = () => {
+  const env = process.env.NODE_ENV || "development";
+  const isDevelopment = env === "development";
+  return isDevelopment ? "debug" : "warn";
```
[the remark this answers is not in the document]
Comment (codesankalp) [reply server/utils/logger.js:15]: It depends on the process like development or builds. But I will add an environment variable to override this also.

File: server/loggerConfig.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (nemesisdesign) [reply server/loggerConfig.js:7]: Make sure that by default everything goes to one file.

File: README.md [hunk 4]
```diff
@@ -325,6 +325,22 @@ login_form:
 
 In order to enable users to log via third-party services like Google and Facebook, the ["Social Login" feature of OpenWISP Radius](https://openwisp-radius.readthedocs.io/en/latest/user/social_login.html) must be configured and enabled.
 
+#### Configuring Logging
+
+There are certain environment variables used to configure server logging. The details of environment variables to configure logging are mentioned below:
+
+| Environment Variable | Detail                                                                                                                                                                                        |
+| -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| **LOG_LEVEL**        | (optional) This can be used to set the level of logging. The available values are `warn` and `debug`. By default log level is set to `debug` during development and `warn` during production. |
```
Comment (nemesisdesign) [inline README.md:334]: > The available values are `warn` and `debug`.

What about `info` and `error`?

**Comment by nemesisdesign** [comment]
Please open the PR from a branch of this repository.

**Pull request closed without merging**